src/clojure/nrepl/cmdline.clj

```diff
@@ -239,64 +249,190 @@
+        {:keys [host port transport]} (connection-opts options)]
+    (merge options
+           {:host host
+            :port port
+            :transport transport
+            :bind (:bind options)
+            :middleware middleware
+            :handler (resolve-handler options middleware)
+            :greeting (resolve-greeting options transport)
+            :ack-port (resolve-ack-port options)})))
+
+(defn connect-to-server
+  "Connects to a running nREPL server and runs a REPL. Exits program when REPL
+  is closed.
+  Takes a map of nREPL CLI options."
+  [options]
+  (let [{:keys [host port transport]} (connection-opts options)]
+    (run-repl host port {:transport transport})
+    (exit 0)))
+
+(defn ack-server!
+  "Acknowledge the port of this server to another nREPL server running on
+  :ack port.
+  Takes nREPL server map and processed CLI options map.
+  Prints a message describing the acknowledgement between servers.
+  Returns nil."
+  [server options]
+  (when-let [ack-port (resolve-ack-port options)]
+    (let [port (:port server)
+          transport (:transport options)]
+      (binding [*out* *err*]
+        (println (format "ack'ing my port %d to other server running on port %d"
+                         port ack-port)
+                 (:status (send-ack port ack-port transport)))))))
+
+(defn connection-header
+  "Returns nREPL server header that some tools rely on to parse the
+  connection details from.
+  Takes nREPL server map and processed CLI options map.
+  Returns connection header string."
+  [server options]
+  (let [transport (:transport options)
+        port (:port server)
+        ^java.net.ServerSocket ssocket (:server-socket server)
+        host (.getHostName (.getInetAddress ssocket))]
+    ;; The format here is important, as some tools (e.g. CIDER) parse the string
+    ;; to extract from it the host and the port to connect to
+    (format "nREPL server started on port %d on host %s - %s://%s:%d"
+            port host (transport/uri-scheme transport) host port)))
+
+(defn save-port-file!
+  "Writes a file relative to project classpath with port number so other tools
+  can infer the nREPL server port.
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  ;; Many clients look for this file to infer the port to connect to
+  (let [port (:port server)
+        port-file (io/file ".nrepl-port")]
+    (.deleteOnExit port-file)
+    (spit port-file port)))
+
+(defn interactive-repl!
+  "Runs an interactive repl if :interactive CLI option is true otherwise
+  puts the current thread to sleep
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  (let [transport (:transport options)
+        host (:host server)
+        port (:port server)]
+    (run-repl host port (merge (when (:color options) colored-output)
+                               {:transport transport}))))
+
+(defn create-server
```

Comment:
As this is nothing but a call of `start-server` I really doubt we need it.

Reply:
Since we're building a public API for the cmdline, it represents a crucial user goal. It was also done for scalability. If the server initialization process needs to deviate, there's a documented API and a place to put that logic :smile:

Reply:
I doubt the API is ever going to change, but the current function already illustrates the problem of having redundant functions in your API - it's docstring is outdated. :-)

Reply:
If we keep the function I guess it should be named again `start-server` as I'd be confused as to whether there's some difference between `start-server` and `create-server`.

Reply:
Hah! That is a very good point 😞 If you can please humor me writing it out:

**If we remove the API function**
- Users will need to import the start-server function directly which means importing from both the nrepl server ns and the cmdline ns.
- Users will possibly need to each write an identical function to the API start-server function to get the values from the options map returned by the `server-opts` function.
- We don't have to worry about extra tests or docs to support it.

**If we keep the API function**
- Users only need to import the required functions from the cmdline ns
- Users can reuse the `start-server` API function
- The `start-server` API function composes nicely with the `server-opts` function.


Unless there's anything you think I missed or got wrong I think the work to update the docs and writing a test would be less overall work than requiring future devs to each have to reimplement a similar function. I prefer to provide meaningful building blocks but I do see your point about the maintenance cost. 

I'll leave this decision to you 🤔

Reply:
I can live with keeping the function.

Reply:
@bbatsov One more question. How would you recommend wrapping the original start-server fn? 

```clj
(ns nrepl.cmdline
  (:require [nrepl.server :refer [start-server] :rename {start-server start-nrepl-server]))

;;...

(defn start-server
  []
  (start-nrepl-server ...))
```

```clj
(ns nrepl.cmdline
  (:require [nrepl.server :as server]))

;; ...

(defn start-server
  []
  (server/start-server ...))
```

Or do you have another way you would recommend?